File: junit-runner/src/main/java/cmu/pasta/fray/runner/JUnitRunner.java
package cmu.pasta.fray.runner;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.lang.reflect.Method;

import org.junit.platform.engine.discovery.DiscoverySelectors;
import org.junit.platform.launcher.Launcher;
import org.junit.platform.launcher.LauncherDiscoveryRequest;
import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder;
import org.junit.platform.launcher.core.LauncherFactory;
import org.junit.platform.launcher.listeners.SummaryGeneratingListener;
import org.junit.runner.Description;
import org.junit.runner.JUnitCore;
import org.junit.runner.Request;
import org.junit.runner.manipulation.Filter;
import org.junit.runner.notification.Failure;
import org.junit.runner.Result;

public class JUnitRunner {
    public static void main(String[] args) throws ClassNotFoundException {
        boolean isJunit4 = args[0].equals("junit4");
        String[] classAndMethod = args[1].split("#");

        if (isJunit4) {
            Request request = Request.method(
                    Class.forName(classAndMethod[0], true, Thread.currentThread().getContextClassLoader()),
                    classAndMethod[1]
            ).filterWith(new Filter() {
                private boolean found = false;

                @Override
                public boolean shouldRun(Description description) {
                    if (found) return false;
                    if (description.getMethodName().equals(classAndMethod[1])) {
                        found = true;
                    }
                    return found;
                }

                @Override
                public String describe() {
                    return classAndMethod[1];
                }
            });

            Result result = new JUnitCore().run(request);
            if (!result.wasSuccessful()) {
                StringBuilder failureReport = new StringBuilder();
                for (Failure failure : result.getFailures()) {
                    failureReport.append("testHeader: ").append(failure.getTestHeader()).append("\n")
                            .append("trace: ").append(failure.getTrace()).append("\n")
                            .append("description: ").append(failure.getDescription()).append("\n");
                }
                System.out.println(failureReport.toString());
                throw new RuntimeException(failureReport.toString());
            }
        } else {
            Class[] parameterTypes = new Class[0];

            String testClassName = classAndMethod[0];
            Class<?> testClass = Class.forName(testClassName, true, Thread.currentThread().getContextClassLoader());
            for (Method method : testClass.getDeclaredMethods()) {
                if (method.getName().equals(classAndMethod[1])) {
                    System.out.println(method.getName());
                    parameterTypes = method.getParameterTypes();
                }
            }
            LauncherDiscoveryRequest request = LauncherDiscoveryRequestBuilder.request()
                    .selectors(DiscoverySelectors.selectMethod(classAndMethod[0], classAndMethod[1], parameterTypes))
                    .build();
            Launcher launcher = LauncherFactory.create();
            SummaryGeneratingListener listener = new SummaryGeneratingListener();
            launcher.registerTestExecutionListeners(listener);
            try {
                launcher.execute(request);
            } catch (Throwable e) {
                e.printStackTrace();
            }
            if (listener.getSummary().getTestsFailedCount() > 0) {
                StringBuilder failureReport = new StringBuilder();
                listener.getSummary().getFailures().forEach(failure -> {
                    StringWriter stringWriter = new StringWriter();
                    PrintWriter writer = new PrintWriter(stringWriter);
                    failure.getException().printStackTrace(writer);
                    failureReport.append("testHeader: ").append(failure.getTestIdentifier()).append("\n")
                            .append("trace: ").append(stringWriter.toString()).append("\n")
                            .append("exception: ").append(failure.getException()).append("\n");
                });
                System.out.println(failureReport.toString());
                throw new RuntimeException(failureReport.toString());
            }
        }
    }
}

File: examples/src/main/kotlin/JUnitRunner.kt
package cmu.pasta.fray.examples

import java.io.PrintWriter
import java.io.StringWriter
import org.junit.platform.engine.discovery.DiscoverySelectors.selectMethod
import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder
import org.junit.platform.launcher.core.LauncherFactory
import org.junit.platform.launcher.listeners.SummaryGeneratingListener
import org.junit.runner.JUnitCore
import org.junit.runner.Request
import org.junit.runner.manipulation.Filter

fun main(args: Array<String>) {
  val isJunit4 = args[0] == "junit4"
  val classAndMethod = args[1].split("#".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

  if (isJunit4) {
    val request =
        Request.method(
                Class.forName(classAndMethod[0], true, Thread.currentThread().contextClassLoader),
                classAndMethod[1],
            )
            .filterWith(
                object : Filter() {
                  var found = false

                  override fun shouldRun(description: org.junit.runner.Description): Boolean {
                    if (found) return false
                    if (description.methodName == classAndMethod[1]) {
                      found = true
                    }
                    return found
                  }

                  override fun describe(): String {
                    return classAndMethod[1]
                  }
                },
            )
    val result = JUnitCore().run(request)
    if (!result.wasSuccessful()) {
      val failureReport =
          result.failures.joinToString {
            "testHeader: ${it.testHeader}\n" +
                "trace: ${it.trace}\n" +
                "description: ${it.description}\n"
          }
      throw RuntimeException(failureReport)
    }
  } else {
    val request =
        LauncherDiscoveryRequestBuilder.request()
            .selectors(selectMethod(classAndMethod[0], classAndMethod[1]))
            .build()
    val launcher = LauncherFactory.create()
    val listener = SummaryGeneratingListener()
    launcher.registerTestExecutionListeners(listener)
    launcher.execute(request)
    val summary = listener.summary
    if (summary.testsFailedCount > 0) {
      val failureReport =
          summary.failures.joinToString {
            val stringWriter = StringWriter()
            val writer = PrintWriter(stringWriter)
            it.exception.printStackTrace(writer)
            "testHeader: ${it.testIdentifier}\n" +
                "trace: $stringWriter\n" +
                "exception: ${it.exception}\n"
          }
      throw RuntimeException(failureReport)
    }
  }
}
